tests/core/CoreTest.py

```diff
@@ -0,0 +1,192 @@
+
+
+class MultiTaskBear(Bear):
+    BEAR_DEPS = set()
+
+    def __init__(self, section, file_dict, tasks_count=1):
+        Bear.__init__(self, section, file_dict)
+        self.tasks_count = tasks_count
+
+    def analyze(self, run_id):
+        return [run_id]
+
+    def generate_tasks(self):
+        # Choose single task parallelization for simplicity. Also use the
+        # section name as a parameter instead of the section itself, as compare
+        # operations on tests do not succeed on them due to the pickling of
+        # multiprocessing to transfer objects to the other process, which
+        # instantiates a new section on each transfer.
+        return (((i,), {}) for i in range(self.tasks_count))
+
+
+class FailingBear(TestBearBase):
+
+    def analyze(self, bear, section_name, file_dict):
+        raise ValueError
+
+
+class CoreTest(unittest.TestCase):
+
+    def setUp(self):
+        logging.getLogger().setLevel(logging.DEBUG)
+
+    @staticmethod
+    def execute_run(bears):
+        results = []
+
+        def on_result(result):
+            results.append(result)
+
+        run(bears, on_result)
+
+        return results
+
+    @staticmethod
+    def get_comparable_results(results):
+        """
+        Transforms an iterable of ``TestResult`` into something comparable.
+
+        Some ``TestResult`` instances returned by ``run`` contain instance
+        values. Due to the ``ProcessPoolExecutor``, objects get pickled,
+        are transferred to the other process and are re-instantiated,
+        effectively changing the id of them. The same happens again on the
+        transfer back in the results, so we need something that can be
+        compared.
+
+        This function extracts relevant values into a tuple, containing::
+
+            (test_result.bear.name,
+             test_result.section_name,
+             test_result.file_dict)
+
+        :param results:
+            The results to transform.
+        :return:
+            A list of comparable results for tests.
+        """
+        return [(result.bear.name, result.section_name, result.file_dict)
+                for result in results]
+
+    def assertTestResultsEqual(self, real, expected):
+        """
+        Test whether results from ``execute_run`` do equal with the ones given.
+
+        This function does a sequence comparison without order, so for example
+        ``[1, 2, 1]`` and ``[2, 1, 1]`` are considered equal.
+
+        :param real:
+            The actual results.
+        :param expected:
+            The expected results.
+        """
+        comparable_real = self.get_comparable_results(real)
+
+        self.assertEqual(len(comparable_real), len(expected))
+        for result in expected:
+            self.assertIn(result, comparable_real)
+            comparable_real.remove(result)
+
+    def test_run_simple(self):
+        # Test single bear without dependencies.
+        section = Section('test-section')
+        filedict = {}
```

Comment:
Sometimes you need two different sections, sometimes not. I wasn't sure how to unify this, so I made variables for them everywhere (seems I missed some).

So shall I use variables consistently or shall I put those sections and file-dicts into `setUp()`?